jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/messages/DispatchedMessageSink.java

```diff
@@ -135,22 +135,28 @@ public void accept(Frame frame, final Callback callback)
             });
         }
 
-        Callback frameCallback = callback;
+        Callback frameCallback;
         if (frame.isFin())
         {
             // This is the final frame we should wait for the frame callback and the dispatched thread.
-            Callback.Completable completableCallback = new Callback.Completable();
-            frameCallback = completableCallback;
-            CompletableFuture.allOf(dispatchComplete, completableCallback).whenComplete((aVoid, throwable) ->
+            Callback.Completable finComplete = Callback.Completable.from(callback);
+            frameCallback = finComplete;
+            CompletableFuture.allOf(dispatchComplete, finComplete).whenComplete((aVoid, throwable) ->
             {
                 typeSink = null;
                 dispatchComplete = null;
-                if (throwable != null)
-                    callback.failed(throwable);
-                else
-                    callback.succeeded();
+                if (throwable == null)
+                    session.demand(1);
             });
         }
+        else
+        {
+            frameCallback = Callback.from(() ->
+            {
+                callback.succeeded();
+                session.demand(1);
+            }, callback::failed);
```

Comment:
We want to preserve the `InvocationType` of `callback`, so do this instead:

```suggestion
            frameCallback = new Callback.Nested(callback)
            {
                @Override
                public void succeeded()
                {
                    super.succeeded();
                    session.demand(1);
                }
            };
```

Reply:
I have changed this usage, but this probably needs a larger review of general `Callback` usage, there is a ton of usages of `from(Runnable success, Consumer<Throwable> failure)` that wrap another callback and don't preserve the `InvocationType`.